File: solutions/r/rna-transcription/rna-transcription.R
# Transcribes a character list representing DNA nucleotides to RNA
to_rna <- function(dna) {
    if (grepl("[^GCTA]", dna)) {
        stop("invalid nucleotide")
    }

    dna_to_rna <- Vectorize(function(nucleotide) {
        switch(nucleotide,
            "G" = "C",
            "C" = "G",
            "T" = "A",
            "A" = "U",
        )
    })

    (dna |> strsplit("") |> unlist()
        |> dna_to_rna() |> paste(collapse = ""))
}

File: solutions/elixir/rna-transcription/lib/rna_transcription.ex
defmodule RnaTranscription do
  @moduledoc false
  @dna_to_rna %{
    ?G => ?C,
    ?C => ?G,
    ?T => ?A,
    ?A => ?U
  }

  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

  iex> RnaTranscription.to_rna('ACTG')
  'UGAC'
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna), do: dna |> Enum.map(&@dna_to_rna[&1])
end
